crates/bevy_ecs/hecs/src/world.rs

```diff
@@ -612,17 +612,35 @@ impl World {
     /// assert_eq!(*world.get::<bool>(e).unwrap(), true);
     /// ```
     pub fn remove<T: Bundle>(&mut self, entity: Entity) -> Result<T, ComponentError> {
+        self.flush();
+        let loc = self.entities.get_mut(entity)?;
+        unsafe {
+            let to_remove = T::with_static_ids(|ids| ids.iter().copied().collect::<HashSet<_>>());
+
+            let old_index = loc.index;
+            let source_arch = &self.archetypes[loc.archetype as usize];
+            let bundle = T::get(|ty, size| source_arch.get_dynamic(ty, size, old_index))?;
+            match self.remove_bundle_internal(entity, to_remove) {
+                Ok(_) => Ok(bundle),
+                Err(err) => Err(err),
+            }
+        }
+    }
+
+    fn remove_bundle_internal<S: core::hash::BuildHasher>(
+        &mut self,
+        entity: Entity,
+        to_remove: std::collections::HashSet<TypeId, S>,
+    ) -> Result<(), ComponentError> {
         use std::collections::hash_map::Entry;
 
-        self.flush();
         let loc = self.entities.get_mut(entity)?;
```

Comment:
Retrieving the location here and in remove() is redundant work. I'd prefer it if we found a way to only do this once (ex: pass the location into remove_bundle_internal)

Reply:
That was not possible, as location needs to be an `&mut` that references `self`, which also needs to be an `&mut` to call `remove_bundle_internal`.

I was able to remove getting the location from `remove` by retrieving the bundle value in `remove_bundle_internal`. When coming from `remove_one_by_one`, this will not return the expected value as we actually can't get it, that's why there is a parameter `check_presence` set to false when coming from this function...